### DISCUSSION TITLE
Exit early on error/failure

### DISCUSSION THREAD

**Discussion by jakubclark** [tox-dev/tox#2425] [Q&A] [answered]
Apologies in advance if this has been asked before, or is documented somewhere, but I haven't managed to find anything.

I have a few environments, to lint, check format, run tests. These are run sequentially. The lint check is run first, followed by the format check and then finally the tests.

Is it possible to have tox exit early, when some command fails? For example, if the lint check results in an error, it would be nice to have tox exit, instead of moving onto the next environments.

**Comment by jugmac00** [comment, marked as answer]
Afaik this it is not supported by tox stop on the first failed environment.

This might be possible with a plugin.

What is your use case?

**Reply by jakubclark** [reply]
Specifically, in CI, it would be nice to exit early, for example when the code format check fails, instead of moving on and running the tests as well.

**Reply by jugmac00** [reply]
Ok, now I got you.

For CI you can use [tox-gh-actions](https://github.com/ymyzk/tox-gh-actions) as following:

tox.ini
https://github.com/jugmac00/flask-reuploaded/blob/307dd201fa1e2943f295ae205e8f7db14488b60a/tox.ini#L56-L62

gh actions configuration
https://github.com/jugmac00/flask-reuploaded/blob/master/.github/workflows/main.yml

This will look like this:
https://github.com/jugmac00/flask-reuploaded/actions/runs/2366430014

All envs nicely separated - and you can configure gh-actions to stop on the first failure.

Well, if you use gh actions :-)

And Hynek has a nice blog post about this:
https://hynek.me/articles/python-github-actions/

**Reply by jakubclark** [reply]
Ah, that's cool, I'll try it out. And yes, I do use GitHub Actions, so this should work for me. Thanks for the quick reply!
